File: LeetCode/112PathSum/PathSum.py
class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        # Inorder traversal -> Time:O(n), Space:O(h), n=numberOfNodes, h=heightOfTree
        def dfs(node, curr_sum):
            if not node:
                return False
            curr_sum += node.val
            if not node.left and not node.right:
                return curr_sum == targetSum
            return dfs(node.left, curr_sum) or dfs(node.right, curr_sum)
        return dfs(root, 0)

File: LeetCode/112PathSum/PathSum.cpp
class Solution {
public:
    bool dfs(TreeNode* node, int sum, int target_sum) {
        // Inorder traversal -> Time:O(n), Space:O(h), n=numberOfNodes, h=heightOfTree
        if (node == NULL) {
            return false;
        }
        int curr_sum = sum + node->val;
        if (node->left == NULL && node->right == NULL) {
            return curr_sum == target_sum;
        }
        return dfs(node->left, curr_sum, target_sum) || dfs(node->right, curr_sum, target_sum);
    }
    bool hasPathSum(TreeNode* root, int targetSum) {
        return dfs(root, 0, targetSum);
    }
};
